File: src/main/java/net/ndrei/teslacorelib/netsync/SimpleNBTMessage.java
package net.ndrei.teslacorelib.netsync;

import io.netty.buffer.ByteBuf;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.common.network.ByteBufUtils;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;

/**
 * Created by CF on 2016-12-03.
 */
@SuppressWarnings("unused")
public class SimpleNBTMessage implements IMessage {
    private NBTTagCompound compound = null;
    private BlockPos pos = null;
    private int dimension;

    @SuppressWarnings("unused")
    public SimpleNBTMessage() { }

    public SimpleNBTMessage(TileEntity entity, NBTTagCompound compound) {
        this(entity.getPos(), entity.getWorld().provider.getDimension(), compound);
    }

    public SimpleNBTMessage(BlockPos pos, int dimension, NBTTagCompound compound) {
        this.pos = pos;
        this.dimension = dimension;
        this.compound = compound;
    }

    public BlockPos getPos() { return this.pos; }
    public int getDimension() { return this.dimension; }
    public NBTTagCompound getCompound() { return this.compound; }

    @Override
    public void fromBytes(ByteBuf buf) {
        NBTTagCompound pos = ByteBufUtils.readTag(buf);
        if (pos != null) {
            this.pos = new BlockPos(pos.getInteger("x"), pos.getInteger("y"), pos.getInteger("z"));
            this.dimension = pos.getInteger("dim");
        }
        this.compound = ByteBufUtils.readTag(buf);
    }

    @Override
    public void toBytes(ByteBuf buf) {
        NBTTagCompound pos = new NBTTagCompound();
        if (this.pos != null) {
            pos.setInteger("x", this.pos.getX());
            pos.setInteger("y", this.pos.getY());
            pos.setInteger("z", this.pos.getZ());
        }
        pos.setInteger("dim", this.dimension);
        ByteBufUtils.writeTag(buf, pos);
        ByteBufUtils.writeTag(buf, (this.compound != null) ? this.compound : new NBTTagCompound());
    }

    public NetworkRegistry.TargetPoint getTargetPoint() {
        if (this.pos != null) {
            return new NetworkRegistry.TargetPoint(this.dimension, this.pos.getX(), this.pos.getY(), this.pos.getZ(), 64);
        }
        return null;
    }
}

File: src/main/java/net/ndrei/teslacorelib/netsync/SimpleNBTMessage.kt
package net.ndrei.teslacorelib.netsync

import io.netty.buffer.ByteBuf
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.ByteBufUtils
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.ndrei.teslacorelib.TeslaCoreLib

/**
 * Created by CF on 2017-06-28.
 */
class SimpleNBTMessage(pos: BlockPos?, dimension: Int?, compound: NBTTagCompound?) : IMessage {
    var compound: NBTTagCompound? = null
        private set
    var pos: BlockPos? = null
        private set
    var dimension: Int? = null
        private set

    @Suppress("unused")
    constructor() : this(null, null, null)

    constructor(entity: TileEntity, compound: NBTTagCompound)
            : this(entity.pos, entity.world.provider.dimension, compound)

    init {
        this.pos = pos
        this.dimension = dimension
        this.compound = compound
    }

    override fun fromBytes(buf: ByteBuf) {
        val pos = ByteBufUtils.readTag(buf)
        if (pos != null) {
            this.pos = BlockPos(pos.getInteger("x"), pos.getInteger("y"), pos.getInteger("z"))
            this.dimension = pos.getInteger("dim")
        } else {
            TeslaCoreLib.logger.warn("Network package received with missing BlockPos information.")
            this.pos = BlockPos(0, 0, 0)
            this.dimension = 0
        }
        this.compound = ByteBufUtils.readTag(buf) ?: NBTTagCompound()
    }

    override fun toBytes(buf: ByteBuf) {
        val pos = NBTTagCompound()
        if (this.pos != null) {
            pos.setInteger("x", this.pos!!.x)
            pos.setInteger("y", this.pos!!.y)
            pos.setInteger("z", this.pos!!.z)
        }
        if (this.dimension != null) {
            pos.setInteger("dim", this.dimension!!)
        }
        ByteBufUtils.writeTag(buf, pos)
        if (this.compound != null) {
            ByteBufUtils.writeTag(buf, this.compound)
        }
    }

    val targetPoint: NetworkRegistry.TargetPoint?
        get() = if ((this.dimension != null) && (this.pos != null))
            NetworkRegistry.TargetPoint(this.dimension!!, this.pos!!.x.toDouble(), this.pos!!.y.toDouble(), this.pos!!.z.toDouble(), 64.0)
        else null
}
